compiler/damlc/tests/src/DA/Test/IncrementalPackageDb.hs

```diff
@@ -0,0 +1,273 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Test.IncrementalPackageDb (main) where
+
+import Control.Monad.Extra
+import DA.Bazel.Runfiles
+import Data.Foldable
+import System.Directory.Extra
+import System.FilePath
+import System.IO.Extra
+import DA.Test.Process
+import Test.Tasty
+import Test.Tasty.HUnit
+import DA.Daml.Options.Packaging.Metadata (metadataFile)
+import Development.IDE.Types.Location
+
+newtype ExpectReinitialization = ExpectReinitialization Bool
+
+main :: IO ()
+main = do
+    damlc <- locateRunfiles (mainWorkspace </> "compiler" </> "damlc" </> exe "damlc")
+    defaultMain $ tests damlc
+
+tests :: FilePath -> TestTree
+tests damlc =
+    testGroup
+        "Incremental package db initialization"
+        [ test
+              "Reinitialize when dependency changes"
+              [ ( "daml.yaml"
+                , unlines
+                      [ "sdk-version: 0.0.0"
+                      , "name: proj"
+                      , "source: daml"
+                      , "version: 0.0.1"
+                      , "dependencies:"
+                      , "  - daml-prim"
+                      , "  - daml-stdlib"
+                      , "  - dependency/dep.dar"
+                      ])
+              , ("daml/A.daml"
+                , unlines
+                      [ "module A where"
+                      , "import B"
+                      , "x : Int"
+                      , "x = f 1"])
+              ]
+              []
+              [ ( "daml.yaml"
+                , unlines
+                      [ "sdk-version: 0.0.0"
+                      , "name: dep"
+                      , "source: daml"
+                      , "version: 0.0.1"
+                      , "dependencies:"
+                      , "  - daml-prim"
+                      , "  - daml-stdlib"
+                      ])
+              , ( "daml/B.daml"
+                , unlines
+                      [ "module B where"
+                      , "template T"
+                      , "  with"
+                      , "    p: Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "f : Int -> Int"
+                      , "f i = i + 1"
+                      ])
+              ]
+              [ ( "daml/B.daml"
+                , unlines
+                      [ "module B where"
+                      , "template T"
+                      , "  with"
+                      , "    p: Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "f : Int -> Int"
```

Comment:
Might be nice to make `A` depend on the change in `B`, e.g., also modify it to reference a new definition that we add in `B`.

Reply:
added an additional reference.